pennylane/capture/meta_type.py

```diff
@@ -0,0 +1,129 @@
+        raise ImportError("Jax is required for plxpr.")
+
+    # TODO: investigate
+    # pennylane/capture/meta_type.py:37:4: W0223: Method 'at_least_vspace' is abstract in class 'AbstractValue' but is not overridden in child class 'AbstractOperator' (abstract-method)
+    # pennylane/capture/meta_type.py:37:4: W0223: Method 'join' is abstract in class 'AbstractValue' but is not overridden in child class 'AbstractOperator' (abstract-method)
+    # pennylane/capture/meta_type.py:37:4: W0223: Method 'update' is abstract in class 'AbstractValue' but is not overridden in child class 'AbstractOperator' (abstract-method)
+
+    class AbstractOperator(jax.core.AbstractValue):
+        """An operator captured into plxpr."""
+
+        def __eq__(self, other):
+            return isinstance(other, AbstractOperator)
+
+        def __hash__(self):
+            return hash("AbstractOperator")
+
+        def _matmul(self, *args):
+            return qml.prod(*args)
+
+        def _mul(self, a, b):
+            return qml.s_prod(b, a)
+
+        def _rmul(self, a, b):
+            return qml.s_prod(b, a)
+
+        def _add(self, a, b):
+            return qml.sum(a, b)
+
+    jax.core.raise_to_shaped_mappings[AbstractOperator] = lambda aval, _: aval
+
+    return AbstractOperator
+
+
+class PLXPRMeta(type):
+    """A metatype that:
+
+    * automatically registers a jax primitive to ``cls._primitive``
+    * Dispatches class creation to ``cls._primitive.bind`` instead of normal class
+    creation when the primitive is defined and plxpr capture is enabled.
+
+    See ``pennylane/capture/explanations.md`` for more detailed information on how this technically
+    works.
+
+    """
+
+    def __init__(cls, *_, **__):
+
+        # Called when constructing a new type that has this metaclass.
+        # Similar to __init_subclass__ , this allows us to run this code
+        # every time we define a new class
+
+        if not has_jax:
+            cls._primitive = None
+            return
+
+        cls._primitive = jax.core.Primitive(cls.__name__)
+
+        @cls._primitive.def_impl
+        def default_call(*args, **kwargs):
+            if "n_wires" not in kwargs:
+                return type.__call__(cls, *args, **kwargs)
+            n_wires = kwargs.pop("n_wires")
+            wires = args[-n_wires:]
+            args = args[:-n_wires]
+            return type.__call__(cls, *args, wires=wires, **kwargs)
+
+        # logic here will be extended when we make more things use this meta class
+        abstract_type = _get_abstract_operator()
+
+        @cls._primitive.def_abstract_eval
+        def abstract_init(*_, **__):
+            return abstract_type()
+
+    def __call__(cls, *args, **kwargs):
+        # this method is called everytime we want to create an instance of the class.
+        # default behavior uses __new__ then __init__
+        # when tracing is enabled, we want to
+
+        if not plxpr_enabled() or cls._primitive is None:
```

Comment:
If `plxpr_enabled is True`, we have a guarantee to have JAX installed. Do we need the second condition, then?

Reply:
I think I actually did to silence my IDE from complaining about "`None` has no method `bind`".